### PR TITLE
Fixes data duplication when Hudi Table contains a replace commit

### PR DESCRIPTION
**Pull request by loustler**
## Description
If the Hudi table contains a replace commit generated by the clustering service, duplicate data is returned.
The exclude logic does not work when there is a REPLACE COMMIT.

## Additional context and related issues

This logic not works

https://github.com/trinodb/trino/blob/ca9b4a9c56d3f8c46ad30df361eb2faabb3e2505/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/table/HudiTableFileSystemView.java#L214-L217

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java [hunk 1]
```diff
@@ -185,6 +185,41 @@ public void testCopyOnWriteTableSelectAfterUpdate()
...
+        try {
+            onHudi().executeQuery(format(
+                    """
+                            ALTER TABLE default.%s
+                            SET SERDEPROPERTIES (
+                                hoodie.parquet.small.file.limit=0,
+                                hoodie.clustering.inline=true,
+                                hoodie.clustering.inline.max.commits=2)""",
+                    tableName));
+            assertThat(onHudi().executeQuery("show TBLPROPERTIES default." + tableName))
```
Comment (electrum) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:204]: I don't think it's necessary to validate the properties here. We can assume that `ALTER TABLE` works correctly.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java [hunk 2]
```diff
@@ -240,6 +275,57 @@ public void testMergeOnReadTableSelectAfterUpdate()
...
+                    """
+                            ALTER TABLE default.%s
+                            SET SERDEPROPERTIES (
+                                hoodie.parquet.small.file.limit=0,
+                                hoodie.compact.inline=true,
+                                hoodie.compact.inline.max.delta.commits=2,
+                                hoodie.clustering.inline=true,
+                                hoodie.clustering.inline.max.commits=2)""",
+                    tableName));
+            assertThat(onHudi().executeQuery("show TBLPROPERTIES default." + tableName))
```
Comment (electrum) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:306]: Same here, validating the properties is not necessary.

**Review by electrum — COMMENTED** [review]
How do the product tests create a `REPLACE` commit? They seem to only run `INSERT` on a non-partitioned table, and don't replace any existing data.

**Comment by loustler** [comment]
@electrum 
A `REPLACE` commit is created by Hudi clustering service.
Hudi clustering service can sets by `hoodie.clustering.inline=true`.
It means runs Hudi clustering service when after write operation is complete.
The `hoodie.clustering.inline.max.commits=2` option is like a threshold for triggering Hudi clustering.

- https://hudi.apache.org/docs/configurations#hoodieclusteringinline
- https://hudi.apache.org/docs/table_management#alter-hoodie-config-options
- https://hudi.apache.org/tech-specs/#reader-expectations

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java (hunk 1 above)
Comment (loustler) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java]: Removed it

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java (hunk 2 above)
Comment (loustler) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java]: Removed it

**Comment by loustler** [comment]
@electrum 
Could you review this PR again?

**Review by codope — COMMENTED** [review]
@loustler Can you please rebase and rerun your test? The issue should be fixed by #18213

**Comment by loustler** [comment]
@codope 
Awesome. I'll close this PR. Thanks

**Pull request closed without merging**